### DISCUSSION TITLE
feature - better connection syntax

### DISCUSSION THREAD

**Discussion by jmarshall9120** [FreeOpcUa/opcua-asyncio#891] [Ideas]
I feel like the connection syntax here is a little off base from most other python libraries that handle similar operations.  Currently we connect in two flavors:

```
import os
from asyncua import Client

# Flavor 1
with Client(os.getenv('OPCUA_URL')) as client:
    # ....do stuff here

# Flavor 2
client = Client(os.getenv('OPCUA_URL'))
client.set_user(os.getenv('OPCUA_USER'))
client.set_password(os.getenv('OPCUA_PASSWORD'))
try:
    await client.connect()
    # ...do stuff here
except Exception as e:
    #...do error handling here
finally:
    try:
        await client.disconnect()
    except:
        # eat potential second error if the connection never opened successfully
        pass
```

A far more conventional approach would be to return a `connection` object which would functionally be what the `client` object currently is.  The `client` object would become a factory that returns the connection and allows access consistently visa vi the `with` operator.  This is what really makes it 'worth it' it terms of python IMO, as it saves lots of boiler plate. 

```
client = Client(os.getenv('OPCUA_URL'))
client.set_user(os.getenv('OPCUA_USER'))
client.set_password(os.getenv('OPCUA_PASSWORD'))
with client.connect() as conn:
    # ....do something
```
Any thoughts toward this?  The implementation has been done in many other packages, so we could just monkey see monkey do.  Compatibility for this sort of thing has also been solved by passing unknown methods down to the `connection` class

```
class Client:
    #.....
    def __getattr__(self, name):
        return getatttr(Connection, name)
```

An example of package that implemented this a little into the lifespan of the package would be SQLAlchemy.

**Comment by oroulet** [comment]
connect does not return a connection but the with and async with syntax is fully supported: https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/client-minimal.py

as for which one is more common or correct I have no ideas.

**Reply by jmarshall9120** [reply]
That example doesn't work with authentication... unless I botched my testing somewhere.  `with`, calls connect so there is no chance to `set_user` or `set_password`.  I could put my password and user in the url, but that is unideal, prone to certain security defeats and bugs.
